### PR TITLE
ISSUE-785 JMAPSettingRepository api interface + Inmemory implement + Test contract and memory tests

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/tmail-backend/issues/785

### CODE REVIEW

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/settings/JmapSettings.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.linagora.tmail.james.jmap.settings.JmapSettings.{JmapSettingsKey, JmapSettingsValue}
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.refineV
+import eu.timepit.refined.string.MatchesRegex
+import org.apache.james.jmap.core.UuidState
+
+object JmapSettings {
+
+  private type SettingKeyRegex = MatchesRegex["^[a-zA-Z0-9-_#.]{1,255}$"]
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/settings/JmapSettings.scala:11]: Please no regexp

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/settings/JmapSettingsRepository.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import reactor.core.scala.publisher.SMono
+
+import scala.collection.mutable
+
+trait JmapSettingsRepository {
+  def get(username: Username): Publisher[JmapSettings]
+
+  def save(username: Username, settings: JmapSettingsUpsertRequest): Publisher[SettingsStateUpdate]
+
+  def reset(username: Username, settings: JmapSettingsUpsertRequest): Publisher[SettingsStateUpdate]
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/settings/JmapSettingsRepository.scala:20]: What is the difference between `save` and `reset` ?

Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/settings/JmapSettingsRepository.scala:20]: reset will remove all redundant settings, save will not

Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/settings/JmapSettingsRepository.scala:20]: And does it make sense, is it usefull? IMO we just need `reset` being exposed by the API

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/settings/JmapSettings.scala [hunk 3]
```diff
@@ -1,19 +1,33 @@
...
-  private type SettingKeyType = String Refined SettingKeyRegex
+  type SettingKeyType = String Refined JmapSettingKeyConstraint
+
+  private val charMatcher: CharMatcher = CharMatcher.inRange('a', 'z')
+    .or(CharMatcher.inRange('0', '9'))
+    .or(CharMatcher.inRange('A', 'Z'))
+    .or(CharMatcher.is('.'))
+    .or(CharMatcher.is('_'))
+    .or(CharMatcher.is('-'))
+    .or(CharMatcher.is('#'))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/settings/JmapSettings.scala:20]: <3

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**